scripts/check-api-schema.ts

```diff
@@ -0,0 +1,137 @@
+ * that we have documented all settings.
+ *
+ * @note This does not check that we declare enums.
+ */
+
+import fs from 'fs';
+
+import _ from 'lodash';
+import yaml from 'yaml';
+
+import { defaultSettings } from '@/config/settings';
+import { RecursiveReadonly } from '@/utils/typeUtils';
+
+const schemaPath = 'src/assets/specs/command-api.yaml';
+
+interface schemaObject {
+  type: 'object';
+  properties?: Record<string, schemaNode>;
+  additionalProperties?: boolean;
+}
+interface schemaString {
+  type: 'string';
+}
+interface schemaInteger {
+  type: 'integer';
+}
+interface schemaBoolean {
+  type: 'boolean';
+}
+interface schemaMissing {
+  // This is not a real schema type; it's a stand-in for a missing property.
+  type: '<missing>';
+}
+type schemaNode = schemaObject | schemaString | schemaInteger | schemaBoolean | schemaMissing;
+
+const blacklistedPaths = [
+  'kubernetes.checkForExistingKimBuilder',
+  'version',
+];
+
+function checkObject(setting: RecursiveReadonly<any>, schema: schemaNode, path: string[], allowMissing = false): string[] {
+  const errors: string[] = [];
+  const pathString = path.join('.');
+
+  function logTypeError(desiredType: schemaNode['type']) {
+    errors.push(`${ pathString } has incorrect type "${ schema.type }", should be "${ desiredType }"`);
+  }
+
+  if (blacklistedPaths.includes(pathString)) {
+    if (schema.type !== '<missing>') {
+      errors.push(`${ pathString } should not be in the schema`);
+    }
+
+    return errors;
+  }
+
+  if (schema.type === '<missing>') {
+    if (!allowMissing) {
+      errors.push(`${ pathString } is missing in the schema`);
+    }
+
+    return errors;
+  }
+
+  switch (typeof setting) {
+  case 'object': {
+    if (schema.type !== 'object') {
+      logTypeError('object');
+      break;
+    }
+    const schemaProps = schema.properties ?? <Record<string, schemaNode>>{};
+
+    for (const prop in setting) {
+      const propSchema = schemaProps[prop] ?? { type: '<missing>' };
+
+      errors.push(...checkObject(setting[prop], propSchema, path.concat(prop), schema.additionalProperties));
+    }
+    for (const prop in schemaProps) {
+      if (!(prop in setting)) {
+        errors.push(`${ path.concat(prop).join('.') } not found in Settings`);
+      }
+    }
+    break;
+  }
+  case 'boolean':
+    if (schema.type !== 'boolean') {
+      logTypeError('boolean');
+    }
+    break;
+  case 'number':
+    if (schema.type !== 'integer') {
+      logTypeError('integer');
+    }
+    break;
+  case 'string':
+    if (schema.type !== 'string') {
+      logTypeError('string');
+    }
+    break;
+  default:
+    errors.push(`${ pathString } has object of unknown type ${ typeof setting }`);
+  }
+
+  return errors;
+}
+
+(async function() {
+  const schema = yaml.parse(await fs.promises.readFile(schemaPath, { encoding: 'utf-8' }));
+  const errors = checkObject(defaultSettings, schema.components.schemas.preferences, []);
```

Comment:
Is there any way to make the last arg (path) a bit clear? right now it's confusing to see `[]`, can it have a default perhaps?

Reply:
Fixed in eedef695d5b48d96d246f88b3959e592723b24cd